File: frontend/src/features/schedule/model/schemas.ts
import { z } from "zod"

/**
 * WHY: Định nghĩa schema cho Ca làm việc từ backend.
 */
export const shiftSchema = z.object({
  id: z.string().uuid(),
  name: z.string().min(1, "Tên ca không được để trống"),
  start_time: z.string(), // ISO time string or HH:mm:ss
  end_time: z.string(),
  color_code: z.string().nullable().optional(),
  created_at: z.string().optional(),
  updated_at: z.string().optional(),
})

export type Shift = z.infer<typeof shiftSchema>

/**
 * WHY: Trạng thái lịch làm việc.
 */
export enum ScheduleStatus {
  DRAFT = "DRAFT",
  PUBLISHED = "PUBLISHED",
  CANCELLED = "CANCELLED",
}

export type ScheduleStatusType = ScheduleStatus


/**
 * WHY: Schema cho lịch làm việc của nhân viên.
 */
export const staffScheduleSchema = z.object({
  id: z.string().uuid(),
  staff_id: z.string().uuid(),
  shift_id: z.string().uuid(),
  work_date: z.string(), // ISO date string YYYY-MM-DD
  status: z.nativeEnum(ScheduleStatus),
  shift: shiftSchema.optional(),
  staff: z.any().optional(), // Sẽ định nghĩa chi tiết sau nếu cần
})

export type StaffSchedule = z.infer<typeof staffScheduleSchema>

/**
 * WHY: Schema để tạo lịch mới.
 */
export const createStaffScheduleSchema = z.object({
  staff_id: z.string().uuid("Cần chọn nhân viên"),
  shift_id: z.string().uuid("Cần chọn ca làm việc"),
  work_date: z.string().regex(/^\d{4}-\d{2}-\d{2}$/, "Ngày không hợp lệ"),
  status: z.nativeEnum(ScheduleStatus),
})

export type CreateStaffSchedule = z.infer<typeof createStaffScheduleSchema>

/**
 * WHY: Schema cho tạo lịch hàng loạt.
 */
export const batchCreateScheduleSchema = z.object({
  staff_id: z.string().uuid("Cần chọn nhân viên"),
  shift_id: z.string().uuid("Cần chọn ca làm việc"),
  work_dates: z.array(z.string().regex(/^\d{4}-\d{2}-\d{2}$/)),
  status: z.nativeEnum(ScheduleStatus),
})

export type BatchCreateSchedule = z.infer<typeof batchCreateScheduleSchema>

File: backend/app/modules/scheduling/schemas.py
"""
Scheduling Schemas - DTOs cho Shifts và StaffSchedules.
"""
from datetime import date, time
from uuid import UUID

from pydantic import ConfigDict, model_validator
from sqlmodel import SQLModel
from typing_extensions import Self

from app.modules.scheduling.models import ScheduleStatus


class ShiftCreate(SQLModel):
    """Schema tạo mới ca làm việc."""
    name: str
    start_time: time
    end_time: time
    color_code: str | None = None

    @model_validator(mode="after")
    def validate_time_range(self) -> Self:
        """Giờ bắt đầu phải trước giờ kết thúc."""
        if self.start_time >= self.end_time:
            raise ValueError("Giờ bắt đầu phải trước giờ kết thúc")
        return self


class ShiftUpdate(SQLModel):
    """Schema cập nhật ca làm việc."""
    name: str | None = None
    start_time: time | None = None
    end_time: time | None = None
    color_code: str | None = None


class ShiftRead(SQLModel):
    """Schema trả về cho client."""
    id: UUID
    name: str
    start_time: time
    end_time: time
    color_code: str | None = None

    model_config = ConfigDict(from_attributes=True)


class StaffScheduleCreate(SQLModel):
    """Schema tạo phân công lịch làm việc."""
    staff_id: UUID
    shift_id: UUID
    work_date: date
    status: ScheduleStatus = ScheduleStatus.DRAFT


class StaffScheduleBatchCreate(SQLModel):
    """Schema tạo hàng loạt phân công lịch."""
    staff_id: UUID
    shift_id: UUID
    work_dates: list[date]
    status: ScheduleStatus = ScheduleStatus.DRAFT


class StaffScheduleBatchDelete(SQLModel):
    """Schema xóa hàng loạt phân công lịch."""
    schedule_ids: list[UUID]


class StaffScheduleUpdate(SQLModel):
    """Schema cập nhật lịch làm việc (chỉ đổi status hoặc shift)."""
    shift_id: UUID | None = None
    status: ScheduleStatus | None = None


class StaffScheduleRead(SQLModel):
    """Schema trả về cho client."""
    id: UUID
    staff_id: UUID
    shift_id: UUID
    work_date: date
    status: ScheduleStatus

    model_config = ConfigDict(from_attributes=True)


class StaffScheduleReadWithDetails(StaffScheduleRead):
    """Schema trả về kèm thông tin chi tiết Shift và Staff."""
    shift_name: str | None = None
    shift_color: str | None = None
    staff_name: str | None = None
